flang/lib/Parser/message.cpp

```diff
@@ -469,7 +486,7 @@ void Messages::Emit(llvm::raw_ostream &o, const AllCookedSources &allCooked,
     }
     msg->Emit(o, allCooked, echoSourceLines, hintFlagPtr);
     lastMsg = msg;
-    if (msg->IsFatal()) {
+    if (msg->IsFatal(warningsAreErrors)) {
```

Comment:
could this be just `warningsAreErrors || msg->IsFatal()` and save the message classes from all the new complexity?

Reply:
It could and I thought about reverting it to that after you had me back out displaying warnings as error messages, but if we decide that only warnings or some subset of messages get promoted then we would need it this way or something similar.

Reply:
If that happens, we'll do something.  But if this works, it avoids a fair amount of redundancy and confusion that currently doesn't add value.